### DISCUSSION TITLE
Improve Nextflow performance on k8s

### DISCUSSION THREAD

**Discussion by namxle** [nextflow-io/nextflow#4547] [Q&A] [answered]
Hello, currently, I am running Nextflow on a Kubernetes cluster and trying to improve its performance. While going through the logs, I noticed that within these steps, it took nearly 10 minutes, whereas running it in a local environment took only a few seconds.

<img width="1030" alt="image" src="https://github.com/nextflow-io/nextflow/assets/103374580/4b5dad4c-7ca2-420a-bf4f-8170e039b45a">

The nextflow version is [23.04.3.](url)

Can anyone help me optimize this or provide any reasons behind this? Thanks a lot.

**Comment by namxle** [comment, marked as answer]
Finally I have been able to solve this problem. The reason for this delay is that: First, I am using the nextflow docker image from docker hub to start the pipeline. Apparently that before nextflow parsing script and stuff, it downloads some dependencies (maybe related with nf-core or my pipeline itself). And also because of my k8s cluster was placed in Aliyun (Alibaba cloud) Container Service, the internet from the pod to outside was very slow, that caused this delay.
So my solution for this was creating a new image from the nextflow docker image and then running the pipeline first so it can download all the stuff. After that pushing the image to any Container Registry and using that image as the base image for running nextflow pipeline in k8s cluster.

**Reply by bentsherman** [reply]
Thanks for letting us know. The Nextflow docker image downloads most of its dependencies, but it does not download any plugins, which is probably what your pipeline tried to download. Your nextflow log should tell you which plugins it used, so in your custom Docker image you could just download those plugins instead of running the entire pipeline:
```bash
nextflow plugins install <name>@<version>
```

Should make the image smaller and build faster

**Reply by namxle** [reply]
Thank you for your advice, @bentsherman. And a huge thanks for your nf-k8s-best-practices, that was very useful for the nextflow deployment into k8s.

**Reply by namxle** [reply]
@bentsherman, may I ask you a question? My company in China is planning to deploy the application locally on a hospital's private aliyun (without internet access). We want to secure the nextflow pipeline so that our clients cannot read it. Do you have any suggestions? Thank you.
